### PR TITLE
Refactor call protocol

### PR DESCRIPTION
**Pull request by mmun**
- Moves blocks into the call protocol
- Adds a new return value register `v0`
- Simplify the stack checks

### CODE REVIEW

File: packages/@glimmer/opcode-compiler/lib/opcode-builder.ts [hunk 1]
```diff
@@ -587,9 +596,26 @@ export abstract class OpcodeBuilder<Specifier> {
...
+
     let count = this.compileParams(params);
 
+    let flags = count << 4;
+
+    if (synthetic) flags |= 0b1000;
+
+    if (blocks) {
+      flags |= 0b111;
+      // if (blocks.main) flags |= 0b0100;
```
Comment (chadhietala) [inline packages/@glimmer/opcode-compiler/lib/opcode-builder.ts]: Remove?

File: packages/@glimmer/opcode-compiler/lib/opcode-builder.ts [hunk 2]
```diff
@@ -733,16 +759,19 @@ export abstract class OpcodeBuilder<Specifier> {
 
     let { symbols } = symbolTable;
 
+    // let blocks = { main: block, else: inverse, attrs };
```
Comment (chadhietala) [inline packages/@glimmer/opcode-compiler/lib/opcode-builder.ts]: Remove

File: packages/@glimmer/runtime/lib/compiled/opcodes/component.ts [hunk 3]
```diff
@@ -116,21 +116,20 @@ APPEND_OPCODES.add(Op.IsComponent, vm => {
...
 
   let meta = vm.constants.getSerializable<TemplateMeta>(_meta);
   let resolver = vm.constants.resolver;
-  let definition = check(stack.pop(), CheckReference);
 
-  stack.push(new CurryComponentReference(definition, resolver, meta, captured));
+  vm.loadValue(Register.v0, new CurryComponentReference(definition, resolver, meta, captured));
 
-  expectStackChange(vm.stack, -args.length - 1, 'CurryComponent');
+  // expectStackChange(vm.stack, -args.length - 1, 'CurryComponent');
```
Comment (chadhietala) [inline packages/@glimmer/runtime/lib/compiled/opcodes/component.ts]: This likely needs to be brought back and incorporate the new `pop` from the stack.

File: packages/@glimmer/opcode-compiler/lib/opcode-builder.ts [hunk 4]
```diff
@@ -587,9 +596,26 @@ export abstract class OpcodeBuilder<Specifier> {
...
+  compileArgs(params: Option<WireFormat.Core.Params>, hash: Option<WireFormat.Core.Hash>, blocks: Option<Blocks>, synthetic: boolean) {
+    if (blocks) {
+      this.pushYieldableBlock(blocks.main);
+      this.pushYieldableBlock(blocks.else);
+      this.pushYieldableBlock(blocks.attrs);
+    }
+
     let count = this.compileParams(params);
 
+    let flags = count << 4;
```
Comment (chadhietala) [inline packages/@glimmer/opcode-compiler/lib/opcode-builder.ts]: I'm a bit worried that the encoding we are introducing is spread between the builder and the encoder. I'm wondering if we should move all the encoding to the encoder so we have a better idea of heap layout and the available and reserved size of different types of operands.

**Comment by mmun** [comment]
Merging as is, but @wycats will circle back on the comments later.

**Pull request merged**